### ISSUE TITLE
Add Fourier checking as a tutorial

### ISSUE THREAD

**Issue by balopat** [quantumlib/Cirq#3687]
**Is your feature request related to a use case or problem? Please describe.**

As described in https://arxiv.org/abs/0910.4698 Fourier Checking is a simple quantum algorithm that can decide whether two 2^n random bitstrings are "forrelated" or not. I would like to showcase with this problem oracle query complexity and oracle synthesis using diagonal gates. 

**Describe the solution you'd like**

We need the following: 
- a docs/tutorials/fourier_checking.ipynb that explains 
   - Fourier Checking and why it's interesting
   - create classically two distributions based on n qubits that are within the promised boundaries to be forrelated 
   - explains query complexity
~- why simple oracles based on truth tables are just simple diagonal gates - and how you can decompose those into 1 and 2 qubit gates in Cirq~
~- we will need an extension of the ThreeQubitDiagonalGate class to multiple qubits. An initial starting point around this would be this paper: https://iopscience.iop.org/article/10.1088/1367-2630/16/3/033040/meta - but if a contributor finds a more efficient algorithm, that would be even better.~

**Additional context (e.g. screenshots)**

I did some hacking on this colab, that could be used as a starting point: https://colab.research.google.com/drive/1iT-PO6GSstreo5edvxdvVQ2Xj4fxHRXo?usp=sharing. 

**What is the urgency from your perspective for this issue? Is it blocking important work?**

P3 - I'm not really blocked by it, it is an idea I'd like to discuss / suggestion based on principle

**Comment by ybc1991** [comment]
I can contribute to this.

**Comment by balopat** [comment]
Thanks @ybc1991 - as discussed, I'll also split out the oracle synthesis into a separate task from this one.

**Comment by ybc1991** [comment]
@balopat Hi Balint, when you have time, would you please create another issue to specify the requirement of oracle synthesis? Thanks.

**Comment by dabacon** [comment]
Interesting paper out showing some easy cases https://scirate.com/arxiv/2102.06963 and connecting them to recursive QAOA.
